### PR TITLE
Fix for `setindex!` with array values

### PR DESCRIPTION
**Pull request by torfjelde**
The current `possible` implementation for `setindex!` only allows scalars as values, and so something like `x[:, 1] = zeros(size(x, 1))` will result in a `Matrix{Any}`.

This of course has significant perf. implications, so it seems worth it to specialize on these cases.

Related: https://github.com/JuliaFolds/BangBang.jl/pull/238 and https://github.com/TuringLang/DynamicPPL.jl/issues/530

### CODE REVIEW

File: src/base.jl [hunk 1]
```diff
@@ -488,6 +488,21 @@ possible(::typeof(_setindex!), ::C, ::V, ::K) where {C <: AbstractDict, V, K} =
...
+# - Indexing with arrays
+function possible(
+    ::typeof(_setindex!),
+    ::C,
+    ::T,
+    ::Vararg
+) where {C <: AbstractArray{<:Real}, T <: AbstractArray{<:Real}}
+    return implements(setindex!, C) &&
+        promote_type(eltype(C), eltype(T)) <: eltype(C)
+end
```
Comment (torfjelde) [inline src/base.jl:504]: I'm not sure if this is the way to go, unfortunately, as it breaks cases such as
```julia
setindex!!([1, 2, 3], [4, 5], 1)
```
which now hits `setindex!`.

File: src/base.jl [hunk 2]
```diff
@@ -481,13 +481,50 @@ Base.@propagate_inbounds _setindex!(xs, v, I...) = (setindex!(xs, v, I...); xs)
...
+    indices::Vararg
+    ) where {M, C <: AbstractArray{<:Real}, T <: AbstractArray{<:Real,M}}
+    return implements(setindex!, C) &&
+        promote_type(eltype(C), eltype(T)) <: eltype(C) &&
+        # This will still return `false` for scenarios such as
+        #
+        #     setindex!!([1, 2, 3], [4, 5, 6], :, 1)
+        #
+        # which are in fact valid. However, this case is rare.
+        (_index_dimension(indices) == M || _index_dimension(indices) == 1)
```
Comment (torfjelde) [inline src/base.jl]: Possibly stupid idea: could we just do `>= M` instead of `== M`?

File: src/base.jl [hunk 3]
```diff
@@ -481,13 +481,50 @@ Base.@propagate_inbounds _setindex!(xs, v, I...) = (setindex!(xs, v, I...); xs)
...
+    indices::Vararg
+    ) where {M, C <: AbstractArray{<:Real}, T <: AbstractArray{<:Real,M}}
+    return implements(setindex!, C) &&
+        promote_type(eltype(C), eltype(T)) <: eltype(C) &&
+        # This will still return `false` for scenarios such as
+        #
+        #     setindex!!([1, 2, 3], [4, 5, 6], :, 1)
+        #
+        # which are in fact valid. However, this cases are rare.
+        (_index_dimension(indices) == M || _index_dimension(indices) == 1)
```
Comment (torfjelde) [inline src/base.jl:525]: Possibly stupid idea: could we just do >= M instead of == M?

**Comment by torfjelde** [comment]
I'd also wager that we could use some functionality from `Base` for all this reasoning about indexes, etc.?

File: src/base.jl (hunk 3 above)
Comment (torfjelde) [inline src/base.jl:525]: For the linear indexing, do we need to also check if `IndexStyle` is linear?

**Comment by torfjelde** [comment]
Would love some feedback on this! @MasonProtter maybe?:) 
In Turing.jl this has quite detrimental performance implications, so it would be nice to get sorted.

File: src/base.jl [hunk 4]
```diff
@@ -481,13 +481,50 @@ Base.@propagate_inbounds _setindex!(xs, v, I...) = (setindex!(xs, v, I...); xs)
...
+_index_dimension(::Colon) = 1
+_index_dimension(::AbstractVector) = 1
+_index_dimension(indices::Tuple) = sum(map(_index_dimension, indices))
+
+function possible(
+    ::typeof(_setindex!),
+    ::C,
+    ::T,
+    indices::Vararg
+    ) where {M, C <: AbstractArray{<:Real}, T <: AbstractArray{<:Real,M}}
```
Comment (MasonProtter) [inline src/base.jl:517]: Why the restrictions to `<:Real` here?

File: src/base.jl (hunk 3 above)
Comment (MasonProtter) [reply src/base.jl:525]: I don't have a good intuition for that. I guess the question would be: can we cook up an example that would be incorrect / error if we did `>=`?

**Review by MasonProtter — COMMENTED** [review]
Hey @torfjelde I'm really sorry for the slow reply. This looks like it's on the right track, but I think there's still some cases that need to be caught, like I think this won't work with
```julia
setindex!([1,2,3,4,5,6], [7,8,9], 1:3)
```
right?

Maybe in `possible`, a better thing to do would be to test if 
```
size(view(collection, indices...)) == size(value)
```
? Do you foresee any problems with that? Or maybe this isn't even `possible`'s job, and we should instead throw an error inside the `setindex!!` impl if they give indices that don't make sense with the size of the value being inserted.

**Comment by MasonProtter** [comment]
@torfjelde what can I do to help get this PR moving again? Is it still causing problems with Turing?

File: src/base.jl (hunk 4 above)
Comment (torfjelde) [reply src/base.jl:517]: I don't remember tbh :confused: Might have been me just wanting to limit the extent of this change to avoid potential bugs.

**Comment by torfjelde** [comment]
> This looks like it's on the right track, but I think there's still some cases that need to be caught, like I think this won't work with

That example does work as it hits the `AbstractVector` dimension-calculation (just checked) :)

> Maybe in possible, a better thing to do would be to test if

I was under the impression that we wanted to avoid runtime checks in `possible`, but it might be worth it in this scenario.

> what can I do to help get this PR moving again? Is it still causing problems with Turing?

Sorry about the delay Mason; we just "addressed" this by type-piracy for now so kind of filed this PR away. But I think the current version is working, it's just a matter of determining the scope of the change, i.e. should it also apply to non-`Real` element types?

**Comment by MasonProtter** [comment]
> That example does work as it hits the AbstractVector dimension-calculation (just checked) :)

huh yeah I'm not sure what I was thinking of then, sorry for the noise.

>  Sorry about the delay Mason; we just "addressed" this by type-piracy for now so kind of filed this PR away. But I think the current version is working, it's just a matter of determining the scope of the change, i.e. should it also apply to non-Real element types?

Not at all a problem, I just wanted to make sure this good work you've done here doesn't get lost. I just checked out the PR again, and tried removing the `<:Real` and everything seems to work, as well as stuff like this:

With the `<:Real` restriction:
```julia
julia> let v = [(1,),(2,),(3,)]
            setindex!!(v, [(4,), (5,), (6,)], :, 1)
            v
       end
3-element Vector{Tuple{Int64}}:
 (1,)
 (2,)
 (3,)
```
without the restriction:
```julia
julia> let v = [(1,),(2,),(3,)]
            setindex!!(v, [(4,), (5,), (6,)], :, 1)
            v
       end
3-element Vector{Tuple{Int64}}:
 (4,)
 (5,)
 (6,)
```

So I say let's just completely drop the element type restriction, maybe add the above block as a test case and merge?

**Comment by MasonProtter** [comment]
test failure is unrelated.

**Comment by torfjelde** [comment]
Wonderful; thanks @MasonProtter !

**Pull request merged**